vervet-underground/internal/storage/mem/mem.go

```diff
@@ -6,81 +6,274 @@ package mem
+			s.serviceVersions[name] = vervet.VersionSlice{*parsedVersion}
+		} else {
+			s.serviceVersions[name] = append(s.serviceVersions[name], *parsedVersion)
+			// sort versions when new ones are introduced to maintain BST functionality
+			sort.Sort(s.serviceVersions[name])
+		}
+	}
+	// End of initializations
+
+	// add the new ContentRevision
+	s.serviceVersionMappedRevisionSpecs[name][version][digest] = ContentRevision{
+		serviceVersion: name + "_" + version,
+		timestamp:      scrapeTime,
+		digest:         digest,
+		blob:           contents,
+	}
+
+	return nil
+}
+
+// CollateVersions does the following:
+//   - calls updateCollatedVersions for a slice of unique vervet.Version entries
+//   - for each unique vervet.Version, run collateVersion to create a compiled VU openapi doc
+func (s *Storage) CollateVersions() error {
+	var errs error
+	s.updateCollatedVersions()
+	for _, version := range s.collatedVersions {
+		err := s.collateVersion(version)
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not Collate Version %s", version.String())
+			errs = multierr.Append(errs, err)
+		}
+	}
+	return errs
+}
+
+func (s *Storage) GetCollatedVersionSpecs() (map[string][]byte, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	versionSpecs := map[string][]byte{}
+	for key, value := range s.collatedVersionedSpecs {
+		json, err := value.MarshalJSON()
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not Marshal Version %s", key.String())
+			return nil, err
+		}
+		versionSpecs[key.String()] = json
+	}
+	return versionSpecs, nil
+}
+
+// updateCollatedVersions collects all unique vervet.Versions detected for every service published using
+func (s *Storage) updateCollatedVersions() {
+	uniqueVersions := make(map[string]vervet.Version)
+
+	// map keys automatically make versions unique rather than iterate through array for each entry
+	// service, versionSlice
+	for _, versionSlice := range s.serviceVersions {
+		// index, version
+		for _, version := range versionSlice {
+			uniqueVersions[version.String()] = version
+		}
+	}
+
+	s.mu.Lock()
+	s.collatedVersions = make([]vervet.Version, len(uniqueVersions))
+	i := 0
+	for _, value := range uniqueVersions {
+		s.collatedVersions[i] = value
+		i++
+	}
+
+	// must sort at end
+	sort.Sort(s.collatedVersions)
+	s.mu.Unlock()
+}
+
+// collateVersion fuzzy matches each service's closest version to the target vervet.Version,
+// collects the latest ContentRevision, if one exists and matches the target.
+// Once all services have been searched, call mergeContentRevisions
+func (s *Storage) collateVersion(version vervet.Version) error {
+	// number of services maximum needed
+	contentRevisions := make([]ContentRevision, 0)
+
+	// preprocessing all relevant docs in byte format before combining
+	for service, versionSlice := range s.serviceVersions {
+		// If there is an exact match on versions 1-to-1
+		if revisions, ok := s.serviceVersionMappedRevisionSpecs[service][version.String()]; ok {
```

Comment:
This will need to be protected by the mutex.
```suggestion
		s.mu.RLock()
		revisions, ok := s.serviceVersionMappedRevisionSpecs[service][version.String()]
		s.mu.RUnlock()
		if ok {
```